src/mpcd.1
.TH MPCD 1 2023-05-22
.SH NAME
\fBmpcd\fR \(en report code clones
.SH SYNOPSIS
\fBmpcd\fR [\fB\-bjSVv\fR] [\fB\-n \fIclone-length\fR]
.SH DESCRIPTION
The \fBmpcd\fR utility reads from its standard input a stream
of file identifiers (e.g. file paths) prefixed with F,
followed by each file's
tokens for each line as integers in the range 0\^\(en\^4294967295 (2\u32\d \(en 1).
It reports on its standard output identified code clones.
Each clone is reported in the following format.
.RS
.TP
\(bu
One line containing the number of clones in the group and
the number of tokens in each group's element.
.TP
\(bu
One line for each clone group element containing
the start line, the end line, and the file identifier.
.TP
\(bu
A blank line.
.RE
All the above elements are tab-separated.


.SH OPTIONS
The behavior of the \fBmpcd\fR utility can be controlled
through the following command-line option.
.RS 3

.TP
.B -b
Identify clone block regions (delimited with \fC{\fP and \fC}\fP),
rather than clone line regions.

.TP
.B -j
Produce JSON rather than plain text output.

.TP
.BI "-n " clone-length
Specify the minimum length of clones that will be detected.
The default value is 15.

.TP
.BI "-S "
Display the program's memory requirements and exit.

.TP
.B -V
Display the program's version number and exit.

.TP
.B -v
Produce verbose output on the standard error with processing details.

.RE

.SH EXAMPLES
.PP
Identify Type 1 (exact) clones in all Java files
located in the \fCsrc\fP directory and output them in JSON format.
The \fItokenizer\fP's
\fC-f\fP option is used to identify each output file and
the \fC-g\fP option to process all identifiers in the same (global) scope.


.ft C
.nf
# Obtain list of Java files in the src directory
find src -type f -name '*.java' |

# Tokenize files in paths coming from stdin, using the same scope for all ids
tokenizer -l Java -o line -i - -g -f |

# Identify clones
mpcd -j >results.json
.ft P
.fi

.PP
Find Type 2 (near) clones in the \fCkernel\fP directory
of the Linux kernel v6.3,
which is stored in a bare Git repository named \fClinux.git\fP.
The \fItokenizer\fP's
\fC-c\fP option is used to output the same token value for all identifiers.
Adjust clone size to 40 tokens and generate verbose output while processing.
The results are saved in textual format.

.ft C
.nf
git --git-dir linux.git ls-tree -r --name-only v6.3 -- kernel |
grep '\.[ch]$' |
while read file; do
  echo "F$file"
  git --git-dir linux.git show "v6.3:$file" |
  tokenizer -l C -o line -c
done |
mpcd -v -n 40 >results.txt
.ft P
.fi

.PP
Identify Type 1 (exact) clones in all JavaScript files
located in the current directory, excluding those in the \fCnode_modules\fP
directory.
Output the results in JSON format.


.ft C
.nf
# Obtain list of required JavaScript files
find . -path './node_modules' -prune -o -type f -name '*.js' |

# Tokenize files in paths coming from stdin, using the same scope for all ids
tokenizer -l JavaScript -o line -i - -g -f |

# Identify clones
mpcd -j >results.json
.ft P
.fi


.SH DIAGNOSTICS
None.

.SH SEE ALSO
.IR tokenizer (1)
\(em convert source code into integer vectors.

.SH AUTHORS
Written by Diomidis Spinellis.

.SH BUGS
For performance reasons no checks are made on the validity of the input
stream.
Invalid input may result in a crash.

Reported clones are regions that have the same tokens up to the specified
length,
extended to the end of the last line,
and then extended as a group line-by-line to cover as many tokens as
possible,
as long as all group members are the same.
No attempt is made to split groups into longer ones covering
differing regions.

Reported clones may overlap.
This may be a feature.


src/mpcd.cpp
/*-
 * Copyright 2023 Diomidis Spinellis
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 *
 * Modular performant clone detector
 *
 */

#include <cstring>
#include <string>
#include <iostream>
#include <ostream>

#include <errno.h>
#include <unistd.h>

#include "TokenContainer.h"
#include "CloneDetector.h"

const char version[] = "1.1.4";

static void
size_report()
{
    // Three pointers per Red-Black tree node plus color overhead
    std::cout << "Bytes per token: " << sizeof(FileData::token_type) << std::endl;
    std::cout << "Bytes per unique line: " << sizeof(SeenTokens) + sizeof(CloneLocation) + 3 * sizeof(void *) + sizeof(int) << std::endl;

    std::cout << "Bytes per duplicate line: " << sizeof(SeenTokens) << std::endl;
    std::cout << "Bytes per file: " << sizeof(FileData) << std::endl;

    std::cout << "Bytes per clone group: " << 2 * sizeof(void *) << std::endl;
    std::cout << "Bytes per clone: " << 2 * sizeof(void *) + sizeof(Clone) << std::endl;
}

// Identify clones among the tokenized input stream
int
main(int argc, char * const argv[])
{
    int opt;
    int clone_tokens = 15; // Minimum number of same tokens to identify a clone
    bool verbose = false;
    bool json = false;
    bool block_regions = false;

    while ((opt = getopt(argc, argv, "bjn:SVv")) != -1)
        switch (opt) {
        case 'b':
            block_regions = true;
            break;
        case 'j':
            json = true;
            break;
        case 'n':
            clone_tokens = std::atoi(optarg);
            if (clone_tokens == 0) {
                std::cerr << "Invalid token number specified" << std::endl;
                exit(EXIT_FAILURE);
            }
            break;
        case 'S':
            size_report();
            exit(EXIT_SUCCESS);
        case 'V':
            std::cout << "mpcd " << version << std::endl;
            exit(EXIT_SUCCESS);
        case 'v':
            verbose = true;
            break;
        default: /* ? */
            std::cerr << "Usage: " << argv[0] <<
                " [-bjSVv] [-n tokens]" << std::endl;
            exit(EXIT_FAILURE);
        }

    if (verbose)
        std::cerr << "Reading input tokens." << std::endl;
    TokenContainer token_container(std::cin);
    if (verbose)
        std::cerr << "Read "
            << token_container.file_size() << " files, "
            << token_container.line_size() << " lines, "
            << token_container.token_size() << " tokens."
            << std::endl;

    CloneDetector cd(token_container, clone_tokens);
    if (verbose)
        std::cerr << "Identified "
            << cd.get_number_of_seen_clones() << " potential clones in "
            << cd.get_number_of_seen_sites() << " total sites."
            << std::endl;

    cd.prune_non_clones();
    if (verbose)
        std::cerr << "Pruned non-clone sites leaving "
            << cd.get_number_of_seen_sites() << " sites."
            << std::endl;

    if (block_regions)
        cd.create_block_region_clones();
    else
        cd.create_line_region_clones();

    cd.clear_clone_candidates();
    if (verbose) {
        std::cerr << "Identified " << cd.get_number_of_clones()
            << " clones in " << cd.get_number_of_clone_groups() << " groups."
            << std::endl;
        if (cd.get_number_of_clone_groups() > 0)
            std::cerr << "Each clone element is on average "
                << cd.get_number_of_clone_tokens() / cd.get_number_of_clone_groups()
                << " tokens long."
                << std::endl;
    }

    if (!block_regions) {
        // Extend line regions as far as possible
        cd.extend_clones();
        if (verbose) {
            std::cerr << "Extended clones to their maximal size." << std::endl;
            if (cd.get_number_of_clone_groups() > 0)
                std::cerr << "Each clone element is on average "
                    << cd.get_number_of_clone_tokens() / cd.get_number_of_clone_groups()
                    << " tokens long."
                    << std::endl;
        }
    }

    cd.remove_shadowed_groups();
    if (verbose)
        std::cerr << "Removed shadowed clone groups, with the result being "
            << cd.get_number_of_clones() << " clones in "
            << cd.get_number_of_clone_groups() << " groups."
            << std::endl;

    if (json)
        cd.report_json();
    else
        cd.report_text();

    exit(EXIT_SUCCESS);
}
